### ISSUE TITLE
Abnormally high scrape duration

### ISSUE THREAD

**Issue by Mawiks** [coreos/prometheus-operator#2749]
## Problem

On our QA cluster, Grafana shows very high scrape durations for Prometheus (the order of magnitude being seconds, as opposed to milliseconds on other environments with identical deploys, metrics and cardinality). Additionally, the scrape duration spikes to the point where Grafana's back-end connection times out and can no longer connect to Prometheus for a while.

![scrape_duration_qa](https://user-images.githubusercontent.com/54933880/64331038-61a81d00-cfd2-11e9-97b2-075c37b33f7f.png)

## Expectation

We have identical setups on a DEV and PROD cluster that do not display these problems.

For example, this is on DEV:

![scrape_duration_dev](https://user-images.githubusercontent.com/54933880/64331057-6a98ee80-cfd2-11e9-9cee-daeee5df6dce.png)

As allocated resources and metrics are near-identical, we expect behavior as in the screenshot above, and no timeouts.

## Reproduce

We have been unable to reproduce this on a different environment, nor have we been able to deploy Prometheus-Operator on this particular cluster without this behavior. We suspect this is an issue with the environment itself, yet all attempts at isolating cluster- or node-specific problems have failed (see below).

## Analysis

Grafana shows an obvious proxy error when this timeout happens:

```
2019/09/04 05:03:48 http: proxy error: context canceled
```

But Prometheus shows there are no problems that seem to point out scraping or resource issues:

```
level=warn ts=2019-09-05T04:56:30.701Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/app/discovery/kubernetes/kubernetes.go:263: watch of *v1.Endpoints ended with: too old resource version: 131248269 (131248660)"
level=warn ts=2019-09-05T04:58:41.192Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/app/discovery/kubernetes/kubernetes.go:263: watch of *v1.Endpoints ended with: too old resource version: 131248034 (131248974)"
level=info ts=2019-09-05T05:03:06.037Z caller=compact.go:499 component=tsdb msg="write block" mint=1567648800000 maxt=1567656000000 ulid=01DKZWPD1F78RGH38R945MV5Q0 duration=3m3.04685161s
level=info ts=2019-09-05T05:03:33.391Z caller=head.go:540 component=tsdb msg="head GC completed" duration=5.089016404s
level=info ts=2019-09-05T05:05:16.013Z caller=head.go:604 component=tsdb msg="WAL checkpoint complete" first=5639 last=5641 duration=1m42.622266155s
level=warn ts=2019-09-05T05:14:43.593Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/app/discovery/kubernetes/kubernetes.go:263: watch of *v1.Endpoints ended with: too old resource version: 131250386 (131251309)"
level=warn ts=2019-09-05T05:16:46.489Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/app/discovery/kubernetes/kubernetes.go:263: watch of *v1.Endpoints ended with: too old resource version: 131251412 (131251606)"
```

Prometheus shows that its endpoints are slow. As an example, here's our AMQ endpoint on the QA cluster:

![prometheus_target_amq_qa](https://user-images.githubusercontent.com/54933880/64331109-7dabbe80-cfd2-11e9-82ea-e2e2c0f278f1.png)

But when we curl these ourselves, only Prometheus-related endpoints are slow, standalone application endpoints are as fast as they are supposed to be despite what Prometheus reports:

```
$ curl -s -o /dev/null -w  "%{time_starttransfer}\n" http://our-amq-endpoint:port/metrics

> 0.287544
```

Yet not a single pod or container is getting killed or showing errors, nor do any of them seem to churn. Prometheus is not starved of resources, nor do any others report excessive use. They are also not limited and are allowed to burst/overcommit. Here is Prometheus itself, for example:

```
λ kubectl -n monitoring get pods -o wide
NAME                                                      READY   STATUS    RESTARTS   AGE     IP               NODE              NOMINATED NODE
 READINESS GATES
...
prometheus-prometheus-operator-prometheus-0               3/3     Running   0          2d      internal_ip      node-0012   <none>
 <none>
...

λ kubectl -n monitoring top pod prometheus-prometheus-operator-prometheus-0
NAME                                          CPU(cores)   MEMORY(bytes)
prometheus-prometheus-operator-prometheus-0   87m          331Mi

λ kubectl top node-0012
NAME              CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%
node-0012         428m         10%    6938Mi          52%
```

## Attempts at fixing the issue

A full purge and reinstall was attempted multiple times, both using version 6.0.0 and version 6.7.3 of the Helm chart, but the issue remains on both.

As we suspected node-related issues, we have tried redeploying both Prometheus and Grafana on different nodes in the cluster. Additionally, we tried deploying them on the same node to rule out inter-node network issues.

As a final overkill, all QA nodes have been redeployed from scratch with Ubuntu 18.04 LTS and re-provisioned with Ansible, identical to how DEV was redeployed a few days prior without problems (Prometheus-Operator worked fine pre- and post-upgrade on DEV).

Kubernetes was upgraded to 1.13.10, no difference.

## Environment

All instances of Prometheus-Operator (QA where the issue occurs, as well as DEV and PROD where it does not), were deployed as follows, on the same day:

```
$ helm delete --purge prometheus-operator
$ kubectl delete crd alertmanagers.monitoring.coreos.com podmonitors.monitoring.coreos.com prometheuses.monitoring.coreos.com prometheusrules.monitoring.coreos.com servicemonitors.monitoring.coreos.com

$ helm install stable/prometheus-operator -f values.yaml -f [values-env.yaml] --name "prometheus-operator" --version 6.0.0 --namespace monitoring
$ helm upgrade prometheus-operator -f values.yaml -f [values-env.yaml] --version 6.7.3 stable/prometheus-operator
```

values.yaml is shared between all environments, values-env.yaml overwrites a few values specifically for that environment. Values-QA is available here:

https://gist.github.com/Mawiks/28664b232b30382518f857cc1a4ff16b

Helm version: 2.14.3
Tiller version: 2.14.3
Helm Chart version: 6.7.3
Prometheus version: 2.9.0
Grafana version: 6.2.5
Kubernetes version: 1.13.10

Has anyone ever encountered something like this? I'm not entirely sure what the underlying issue actually is, let alone where to look next.

Please let me know if should rather report this on the Prometheus or Helm side.

**Comment by brancz** [comment]
Thanks for the thorough write up and investigation. Do you happen to have instrumentation on the endpoints that seem slow to prometheus? That could reveal if only certain scrapes are slower than others. Otherwise does this cluster also have the same amount of resources available? Maybe the overlay network is saturated or noisy neighbors?

**Comment by Mawiks** [comment]
I forced the Prometheus pod to request at least 2000m of CPU to see what happens. Scrape duration instantly went down to an average of 0.025s and no more timeouts have happened now for a while. After some digging I found out there was a limitrange resource on the QA cluster that limited CPU to 100m whereas the other clusters have no such limit and thus allow bursting. As this looks like some legacy configuration that should not have been there, this totally went over my head.

While Prometheus did behave as if it was resource starved, this was not obvious to deduct from the general resource usage. The process did not seem to constantly use the full 100m, often dipping slightly below. Even now it is hovering between 80m and 100m usage, sometimes slightly spiking to 120m. Yet being able to go above 100m has such a massive impact on performance, apparently. I'm not entirely sure how, but at least the core problem is resolved. Scrape duration still spikes (from 0.025s to 0.100s, which the other environments don't do), but this is down orders of magnitude and might be an unrelated issue.

In short, most definitely a problem with the configuration of our cluster, not Prometheus-Operator.
